### PR TITLE
mgr/cephadm: allow RGWSpec networks list to select an IP to bind to

### PR DESCRIPTION
**Pull request by liewegas**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/python-common/ceph/deployment/service_spec.py [hunk 1]
```diff
@@ -706,6 +711,7 @@ def __init__(self,
                  rgw_frontend_type: Optional[str] = None,
                  unmanaged: bool = False,
                  ssl: bool = False,
+                 networks: Optional[List[str]] = None,
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/service_spec.py:714]: I think we have to make this a real property of the base class `ServiceSpec`. We need to be able to make the daemons bind to specific IPs not just for RGWs, but for all other services as well.

I think we can start implementing this for RGWs only, but I think we should attach `netowrks` to the base class already.

File: src/pybind/mgr/cephadm/schedule.py [hunk 2]
```diff
@@ -231,6 +233,26 @@ def get_candidates(self) -> List[DaemonPlacement]:
...
+                    if ips:
+                        ip = sorted(ips)[0]
+                        break
+                if ip:
+                    ls.append(DaemonPlacement(hostname=p.hostname, network=p.network,
+                                              name=p.name, port=p.port, ip=ip))
+                else:
+                    logger.debug(
+                        f'Skipping {p.hostname} with no IP in network(s) {networks}'
+                    )
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/schedule.py:254]: can you move this to a new method?

File: src/python-common/ceph/deployment/service_spec.py (hunk 1 above)
Comment (liewegas) [reply src/python-common/ceph/deployment/service_spec.py:714]: Hmm.  Thinking about this a bit more, there is an argument to be made that this belongs in PlacementSpec.

- It's closer to the HostPlacementSpec, which also has a networks field.  The handling for the two could be converged...
- It consolidates the input for the scheduling in PlacementSpec, making HostAssignment a bit more modular.
- We have to filter the candidate hosts by this when scheduling

The same would apply to port, though.  Which makes it weirder to have default ports associated with service types.

**Review by sebastian-philipp — APPROVED** [review]

**Comment by liewegas** [comment]
jenkins test signed

**Pull request merged**